File: skills/clipboard.py
import asyncio
import logging

import pyperclip

from skills.base_skill import BaseSkill

logger = logging.getLogger(__name__)


class ClipboardSkill(BaseSkill):
    @property
    def name(self) -> str:
        return "clipboard"

    @property
    def description(self) -> str:
        return (
            "Liest oder schreibt Text in die Zwischenablage des Systems. "
            "Kann den aktuellen Inhalt lesen oder neuen Text setzen."
        )

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "enum": ["read", "write"],
                    "description": "'read' liest den Inhalt der Zwischenablage, 'write' setzt neuen Inhalt.",
                },
                "text": {
                    "type": "string",
                    "description": "Der Text, der in die Zwischenablage geschrieben werden soll. Nur bei action='write'.",
                },
            },
            "required": ["action"],
        }

    async def execute(self, action: str = "read", **kwargs) -> str:
        try:
            loop = asyncio.get_event_loop()

            if action == "write":
                text = kwargs.get("text", "")
                if not text:
                    return "Fehler: Kein Text zum Schreiben angegeben."
                await loop.run_in_executor(None, pyperclip.copy, text)
                return f"Text in Zwischenablage geschrieben ({len(text)} Zeichen)."

            # read
            content = await loop.run_in_executor(None, pyperclip.paste)
            if not content:
                return "Die Zwischenablage ist leer."
            # Limit output to avoid overwhelming LLM context
            if len(content) > 5000:
                return f"Zwischenablage-Inhalt ({len(content)} Zeichen, gekuerzt):\n\n{content[:5000]}..."
            return f"Zwischenablage-Inhalt:\n\n{content}"
        except Exception as e:
            logger.exception("Clipboard operation failed")
            return f"Fehler bei Zwischenablage-Operation: {e}"

File: src/tools/builtin/clipboard.ts
import { execFile } from 'node:child_process'
import { promisify } from 'node:util'
import type { Tool } from '../registry.js'

const execFileAsync = promisify(execFile)

export class ClipboardTool implements Tool {
  readonly name = 'clipboard'
  readonly description = 'Liest und schreibt den Inhalt der Zwischenablage.'
  readonly parameters = {
    type: 'object',
    properties: {
      action: {
        type: 'string',
        enum: ['read', 'write'],
        description: 'read = Inhalt lesen, write = Inhalt schreiben',
      },
      text: { type: 'string', description: 'Text zum Schreiben (nur bei action=write)' },
    },
    required: ['action'],
  }

  async execute(args: Record<string, unknown>): Promise<string> {
    const action = String(args.action ?? '')

    try {
      if (process.platform === 'win32') {
        if (action === 'read') {
          const { stdout } = await execFileAsync('powershell', ['-Command', 'Get-Clipboard'], { timeout: 5_000 })
          return stdout.trim() || '(Zwischenablage ist leer)'
        } else {
          const text = String(args.text ?? '')
          await execFileAsync('powershell', ['-Command', `Set-Clipboard -Value '${text.replace(/'/g, "''")}'`], { timeout: 5_000 })
          return `Zwischenablage wurde gesetzt.`
        }
      } else {
        if (action === 'read') {
          try {
            const { stdout } = await execFileAsync('xclip', ['-selection', 'clipboard', '-o'], { timeout: 5_000 })
            return stdout.trim() || '(Zwischenablage ist leer)'
          } catch {
            const { stdout } = await execFileAsync('xsel', ['--clipboard', '--output'], { timeout: 5_000 })
            return stdout.trim() || '(Zwischenablage ist leer)'
          }
        } else {
          const text = String(args.text ?? '')
          try {
            const child = execFile('xclip', ['-selection', 'clipboard'], { timeout: 5_000 })
            child.stdin?.write(text)
            child.stdin?.end()
            await new Promise<void>((res, rej) => child.on('close', c => c === 0 ? res() : rej(new Error(`Exit ${c}`))))
          } catch {
            const child = execFile('xsel', ['--clipboard', '--input'], { timeout: 5_000 })
            child.stdin?.write(text)
            child.stdin?.end()
            await new Promise<void>((res, rej) => child.on('close', c => c === 0 ? res() : rej(new Error(`Exit ${c}`))))
          }
          return 'Zwischenablage wurde gesetzt.'
        }
      }
    } catch (err) {
      return `Fehler: ${err instanceof Error ? err.message : String(err)}`
    }
  }
}
